### DISCUSSION TITLE
Could Integrating External Layout Systems Work?

### DISCUSSION THREAD

**Discussion by zicklag** [RAUI-labs/raui#63] [Q&A]
RAUI's `WidgetUnit` has a specific set of different widget types in it:

```rust
pub enum WidgetUnit {
    None,
    AreaBox(AreaBox),
    PortalBox(PortalBox),
    ContentBox(ContentBox),
    FlexBox(FlexBox),
    GridBox(GridBox),
    SizeBox(SizeBox),
    ImageBox(ImageBox),
    TextBox(TextBox),
}
```

A few of those define the different container types that each use a certain layout technique and properties to lay out their contents. In RAUI you are allowed to make your own layout engine instead of using the default one, but you are still restricted to using the pre-defined `WidgetUnit`s and you can't invent your own layout node type.

I was curious whether or not it was possible to make a design where I could, for instance, add a new kind of layout node that uses something like the [cassowary](https://github.com/dylanede/cassowary-rs) algorithm to lay out children.

I'm not versed in the difficulties or nuances of the UI design or layout process, but would it be possible to put the responsibility for layout on the widget components themselves, and then have a simpler non-layout-specific `WidgetUnit` type that just contains the size and offsets of all ofthe widget types?

I'm not sure if that's just a bad idea and I've got no info to support that that would work, I was just wondering about it because I wanted to try the cassowary algorithm and I wasn't sure if that would require adding a `CassowaryBox` to RAUI core or something like that.

**Comment by PsichiX** [comment]
i was thinking about this for some time - i was wondering how we can modify this set of predefined unit boxes to allow custom plugins make totally new units that doesn't have to or even can't follow settings of the default units (even with breaking changes - this is inevitable to solve this problem). i haven't designed a possible solution yet, but that's gonna be the goal at some point (i would like to allow using RAUI as terminal UI in the future).
i would welcome a general discussion about how we could design such possibility.
